File: core/src/ten_runtime/binding/go/interface/ten_runtime/value_buffer.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//

package ten_runtime

import (
	"encoding/binary"
	"fmt"
	"math"
)

// Buffer protocol constants - must match C layer
const (
	valueBufferMagic      uint16 = 0x10E
	valueBufferVersion    uint8  = 1
	valueBufferHeaderSize        = 8 // magic(2) + version(1) + type(1) + size(4)
)

// Buffer type constants - must match TEN_VALUE_BUFFER_TYPE in C
const (
	bufferTypeInvalid    uint8 = 0
	bufferTypeBool       uint8 = 1
	bufferTypeInt8       uint8 = 2
	bufferTypeInt16      uint8 = 3
	bufferTypeInt32      uint8 = 4
	bufferTypeInt64      uint8 = 5
	bufferTypeUint8      uint8 = 6
	bufferTypeUint16     uint8 = 7
	bufferTypeUint32     uint8 = 8
	bufferTypeUint64     uint8 = 9
	bufferTypeFloat32    uint8 = 10
	bufferTypeFloat64    uint8 = 11
	bufferTypeString     uint8 = 12
	bufferTypeBytes      uint8 = 13
	bufferTypeArray      uint8 = 14
	bufferTypeObject     uint8 = 15
	bufferTypePtr        uint8 = 16
	bufferTypeJSONString uint8 = 17
)

// valueBufferHeader represents the buffer header structure
type valueBufferHeader struct {
	magic    uint16
	version  uint8
	typeName uint8
	size     uint32
}

// newBufferProtocolError creates a new TenError for buffer protocol errors
func newBufferProtocolError(code TenErrorCode, message string) *TenError {
	return NewTenError(code, message)
}

// valueTypeToBufferType converts ValueType to buffer type
func valueTypeToBufferType(vt ValueType) uint8 {
	switch vt {
	case ValueTypeBool:
		return bufferTypeBool
	case ValueTypeInt8:
		return bufferTypeInt8
	case ValueTypeInt16:
		return bufferTypeInt16
	case ValueTypeInt32:
		return bufferTypeInt32
	case ValueTypeInt64:
		return bufferTypeInt64
	case ValueTypeUint8:
		return bufferTypeUint8
	case ValueTypeUint16:
		return bufferTypeUint16
	case ValueTypeUint32:
		return bufferTypeUint32
	case ValueTypeUint64:
		return bufferTypeUint64
	case ValueTypeFloat32:
		return bufferTypeFloat32
	case ValueTypeFloat64:
		return bufferTypeFloat64
	case ValueTypeString:
		return bufferTypeString
	case ValueTypeBytes:
		return bufferTypeBytes
	case ValueTypeArray:
		return bufferTypeArray
	case ValueTypeObject:
		return bufferTypeObject
	case ValueTypePtr:
		return bufferTypePtr
	case ValueTypeJSONString:
		return bufferTypeJSONString
	default:
		return bufferTypeInvalid
	}
}

// bufferTypeToValueType converts buffer type to ValueType
func bufferTypeToValueType(bt uint8) ValueType {
	switch bt {
	case bufferTypeBool:
		return ValueTypeBool
	case bufferTypeInt8:
		return ValueTypeInt8
	case bufferTypeInt16:
		return ValueTypeInt16
	case bufferTypeInt32:
		return ValueTypeInt32
	case bufferTypeInt64:
		return ValueTypeInt64
	case bufferTypeUint8:
		return ValueTypeUint8
	case bufferTypeUint16:
		return ValueTypeUint16
	case bufferTypeUint32:
		return ValueTypeUint32
	case bufferTypeUint64:
		return ValueTypeUint64
	case bufferTypeFloat32:
		return ValueTypeFloat32
	case bufferTypeFloat64:
		return ValueTypeFloat64
	case bufferTypeString:
		return ValueTypeString
	case bufferTypeBytes:
		return ValueTypeBytes
	case bufferTypeArray:
		return ValueTypeArray
	case bufferTypeObject:
		return ValueTypeObject
	case bufferTypePtr:
		return ValueTypePtr
	case bufferTypeJSONString:
		return ValueTypeJSONString
	default:
		return valueTypeInvalid
	}
}

// calculateSerializeSize calculates the total size needed for serialization
func (v *Value) calculateSerializeSize() (int, error) {
	contentSize, err := v.calculateContentSize()
	if err != nil {
		return 0, err
	}
	return valueBufferHeaderSize + contentSize, nil
}

// calculateContentSize calculates the size of the value content
func (v *Value) calculateContentSize() (int, error) {
	switch v.typ {
	case valueTypeInvalid:
		return 0, NewTenError(
			ErrorCodeInvalidType,
			"unsupported value type for size calculation",
		)
	case ValueTypeBool, ValueTypeInt8, ValueTypeUint8:
		return 1, nil
	case ValueTypeInt16, ValueTypeUint16:
		return 2, nil
	case ValueTypeInt32, ValueTypeUint32:
		return 4, nil
	case ValueTypeInt64, ValueTypeUint64:
		return 8, nil
	case ValueTypeFloat32:
		return 4, nil
	case ValueTypeFloat64:
		return 8, nil
	case ValueTypeString:
		str, err := v.GetString()
		if err != nil {
			return 0, fmt.Errorf("failed to get string value: %w", err)
		}
		return 4 + len(str), nil // length(4) + data
	case ValueTypeJSONString:
		str, err := v.GetJSONString()
		if err != nil {
			return 0, fmt.Errorf("failed to get string value: %w", err)
		}
		return 4 + len(str), nil // length(4) + data
	case ValueTypeBytes:
		bytes, err := v.GetBuf()
		if err != nil {
			return 0, fmt.Errorf("failed to get bytes value: %w", err)
		}
		return 4 + len(bytes), nil // length(4) + data

	case ValueTypeArray:
		arr, err := v.GetArray()
		if err != nil {
			return 0, fmt.Errorf("failed to get array value: %w", err)
		}
		size := 4 // array length
		for _, item := range arr {
			size++ // item type
			itemSize, err := item.calculateContentSize()
			if err != nil {
				return 0, fmt.Errorf(
					"failed to calculate array item size: %w",
					err,
				)
			}
			size += itemSize
		}
		return size, nil

	case ValueTypeObject:
		obj, err := v.GetObject()
		if err != nil {
			return 0, fmt.Errorf("failed to get object value: %w", err)
		}
		size := 4 // object size
		for key, val := range obj {
			size += 4 + len(key) // key length + key data
			size++               // value type
			valSize, err := val.calculateContentSize()
			if err != nil {
				return 0, fmt.Errorf(
					"failed to calculate object value size: %w",
					err,
				)
			}
			size += valSize
		}
		return size, nil

	default:
		return 0, NewTenError(
			ErrorCodeInvalidType,
			"unknown value type for size calculation",
		)
	}
}

// serializeToBuffer serializes the Value to a buffer using only Go operations
func (v *Value) serializeToBuffer() ([]byte, error) {
	totalSize, sizeErr := v.calculateSerializeSize()
	if sizeErr != nil {
		return nil, fmt.Errorf(
			"failed to calculate serialize size: %w",
			sizeErr,
		)
	}
	buffer := make([]byte, totalSize)

	pos := 0

	// Write header
	header := valueBufferHeader{
		magic:    valueBufferMagic,
		version:  valueBufferVersion,
		typeName: valueTypeToBufferType(v.typ),
		size:     uint32(totalSize - valueBufferHeaderSize),
	}

	binary.LittleEndian.PutUint16(buffer[pos:], header.magic)
	pos += 2
	buffer[pos] = header.version
	pos++
	buffer[pos] = header.typeName
	pos++
	binary.LittleEndian.PutUint32(buffer[pos:], header.size)
	pos += 4

	// Write content
	err := v.serializeContent(buffer, &pos)
	if err != nil {
		return nil, err
	}

	return buffer, nil
}

// serializeContent serializes the value content to buffer
func (v *Value) serializeContent(buffer []byte, pos *int) error {
	switch v.typ {
	case valueTypeInvalid:
		panic("unsupported value type for serialization")

	case ValueTypeBool:
		boolVal, err := v.GetBool()
		if err != nil {
			return fmt.Errorf("failed to get bool value: %w", err)
		}
		val := uint8(0)
		if boolVal {
			val = 1
		}
		buffer[*pos] = val
		*pos++

	case ValueTypeInt8:
		int8Val, err := v.GetInt8()
		if err != nil {
			return fmt.Errorf("failed to get int8 value: %w", err)
		}
		buffer[*pos] = uint8(int8Val)
		*pos++

	case ValueTypeInt16:
		int16Val, err := v.GetInt16()
		if err != nil {
			return fmt.Errorf("failed to get int16 value: %w", err)
		}
		binary.LittleEndian.PutUint16(buffer[*pos:], uint16(int16Val))
		*pos += 2

	case ValueTypeInt32:
		int32Val, err := v.GetInt32()
		if err != nil {
			return fmt.Errorf("failed to get int32 value: %w", err)
		}
		binary.LittleEndian.PutUint32(buffer[*pos:], uint32(int32Val))
		*pos += 4

	case ValueTypeInt64:
		int64Val, err := v.GetInt64()
		if err != nil {
			return fmt.Errorf("failed to get int64 value: %w", err)
		}
		binary.LittleEndian.PutUint64(buffer[*pos:], uint64(int64Val))
		*pos += 8

	case ValueTypeUint8:
		uint8Val, err := v.GetUint8()
		if err != nil {
			return fmt.Errorf("failed to get uint8 value: %w", err)
		}
		buffer[*pos] = uint8Val
		*pos++

	case ValueTypeUint16:
		uint16Val, err := v.GetUint16()
		if err != nil {
			return fmt.Errorf("failed to get uint16 value: %w", err)
		}
		binary.LittleEndian.PutUint16(buffer[*pos:], uint16Val)
		*pos += 2

	case ValueTypeUint32:
		uint32Val, err := v.GetUint32()
		if err != nil {
			return fmt.Errorf("failed to get uint32 value: %w", err)
		}
		binary.LittleEndian.PutUint32(buffer[*pos:], uint32Val)
		*pos += 4

	case ValueTypeUint64:
		uint64Val, err := v.GetUint64()
		if err != nil {
			return fmt.Errorf("failed to get uint64 value: %w", err)
		}
		binary.LittleEndian.PutUint64(buffer[*pos:], uint64Val)
		*pos += 8

	case ValueTypeFloat32:
		float32Val, err := v.GetFloat32()
		if err != nil {
			return fmt.Errorf("failed to get float32 value: %w", err)
		}
		binary.LittleEndian.PutUint32(
			buffer[*pos:],
			math.Float32bits(float32Val),
		)
		*pos += 4

	case ValueTypeFloat64:
		float64Val, err := v.GetFloat64()
		if err != nil {
			return fmt.Errorf("failed to get float64 value: %w", err)
		}
		binary.LittleEndian.PutUint64(
			buffer[*pos:],
			math.Float64bits(float64Val),
		)
		*pos += 8

	case ValueTypeString, ValueTypeJSONString:
		var stringVal string
		var err error
		if v.typ == ValueTypeString {
			stringVal, err = v.GetString()
		} else {
			stringVal, err = v.GetJSONString()
		}
		if err != nil {
			return fmt.Errorf("failed to get string value: %w", err)
		}

		strBytes := []byte(stringVal)
		strLen := uint32(len(strBytes))

		binary.LittleEndian.PutUint32(buffer[*pos:], strLen)
		*pos += 4

		if strLen > 0 {
			copy(buffer[*pos:], strBytes)
			*pos += int(strLen)
		}

	case ValueTypeBytes:
		bytesVal, err := v.GetBuf()
		if err != nil {
			return fmt.Errorf("failed to get bytes value: %w", err)
		}

		bufLen := uint32(len(bytesVal))

		binary.LittleEndian.PutUint32(buffer[*pos:], bufLen)
		*pos += 4

		if bufLen > 0 {
			copy(buffer[*pos:], bytesVal)
			*pos += int(bufLen)
		}

	case ValueTypeArray:
		arrayVal, err := v.GetArray()
		if err != nil {
			return fmt.Errorf("failed to get array value: %w", err)
		}

		arrayLen := uint32(len(arrayVal))
		binary.LittleEndian.PutUint32(buffer[*pos:], arrayLen)
		*pos += 4

		for _, item := range arrayVal {
			buffer[*pos] = valueTypeToBufferType(item.typ)
			*pos++

			err := item.serializeContent(buffer, pos)
			if err != nil {
				return err
			}
		}

	case ValueTypeObject:
		objectVal, err := v.GetObject()
		if err != nil {
			return fmt.Errorf("failed to get object value: %w", err)
		}

		objSize := uint32(len(objectVal))
		binary.LittleEndian.PutUint32(buffer[*pos:], objSize)
		*pos += 4

		for key, val := range objectVal {
			// Write key
			keyBytes := []byte(key)
			keyLen := uint32(len(keyBytes))

			binary.LittleEndian.PutUint32(buffer[*pos:], keyLen)
			*pos += 4
			copy(buffer[*pos:], keyBytes)
			*pos += int(keyLen)

			// Write value type and content
			buffer[*pos] = valueTypeToBufferType(val.typ)
			*pos++

			err := val.serializeContent(buffer, pos)
			if err != nil {
				return err
			}
		}

	default:
		panic("unsupported value type for serialization")
	}

	return nil
}

// validateBufferHeader validates the buffer header and returns header info
func validateBufferHeader(buffer []byte) (*valueBufferHeader, error) {
	if len(buffer) < valueBufferHeaderSize {
		panic("buffer too small to contain header")
	}

	header := &valueBufferHeader{}
	pos := 0

	header.magic = binary.LittleEndian.Uint16(buffer[pos:])
	pos += 2
	header.version = buffer[pos]
	pos++
	header.typeName = buffer[pos]
	pos++
	header.size = binary.LittleEndian.Uint32(buffer[pos:])

	if header.magic != valueBufferMagic {
		return nil, newBufferProtocolError(ErrorCodeGeneric,
			"invalid buffer magic number")
	}

	if header.version != valueBufferVersion {
		return nil, newBufferProtocolError(ErrorCodeGeneric,
			"unsupported buffer protocol version")
	}

	if bufferTypeToValueType(header.typeName) == valueTypeInvalid {
		return nil, newBufferProtocolError(ErrorCodeGeneric,
			"invalid or unknown buffer type")
	}

	if uint32(len(buffer)) < uint32(valueBufferHeaderSize)+header.size {
		return nil, newBufferProtocolError(ErrorCodeGeneric,
			"buffer size doesn't match header specification")
	}

	return header, nil
}

// deserializeFromBuffer deserializes a Value from buffer
func deserializeFromBuffer(buffer []byte) (*Value, error) {
	header, err := validateBufferHeader(buffer)
	if err != nil {
		return nil, err
	}

	valueType := bufferTypeToValueType(header.typeName)
	pos := valueBufferHeaderSize

	value := &Value{typ: valueType}

	err = value.deserializeContent(buffer, &pos)
	if err != nil {
		return nil, err
	}

	return value, nil
}

// deserializeContent deserializes the value content from buffer
func (v *Value) deserializeContent(buffer []byte, pos *int) error {
	switch v.typ {
	case valueTypeInvalid:
		panic("unsupported value type for deserialization")

	case ValueTypeBool:
		boolVal := buffer[*pos] != 0
		v.data = boolVal
		*pos++

	case ValueTypeInt8:
		int8Val := int8(buffer[*pos])
		v.data = int8Val
		*pos++

	case ValueTypeInt16:
		int16Val := int16(binary.LittleEndian.Uint16(buffer[*pos:]))
		v.data = int16Val
		*pos += 2

	case ValueTypeInt32:
		int32Val := int32(binary.LittleEndian.Uint32(buffer[*pos:]))
		v.data = int32Val
		*pos += 4

	case ValueTypeInt64:
		int64Val := int64(binary.LittleEndian.Uint64(buffer[*pos:]))
		v.data = int64Val
		*pos += 8

	case ValueTypeUint8:
		uint8Val := buffer[*pos]
		v.data = uint8Val
		*pos++

	case ValueTypeUint16:
		uint16Val := binary.LittleEndian.Uint16(buffer[*pos:])
		v.data = uint16Val
		*pos += 2

	case ValueTypeUint32:
		uint32Val := binary.LittleEndian.Uint32(buffer[*pos:])
		v.data = uint32Val
		*pos += 4

	case ValueTypeUint64:
		uint64Val := binary.LittleEndian.Uint64(buffer[*pos:])
		v.data = uint64Val
		*pos += 8

	case ValueTypeFloat32:
		float32Val := math.Float32frombits(
			binary.LittleEndian.Uint32(buffer[*pos:]),
		)
		v.data = float32Val
		*pos += 4

	case ValueTypeFloat64:
		float64Val := math.Float64frombits(
			binary.LittleEndian.Uint64(buffer[*pos:]),
		)
		v.data = float64Val
		*pos += 8

	case ValueTypeString, ValueTypeJSONString:
		strLen := binary.LittleEndian.Uint32(buffer[*pos:])
		*pos += 4

		if strLen == 0 {
			v.data = ""
		} else {
			if *pos+int(strLen) > len(buffer) {
				return newBufferProtocolError(ErrorCodeGeneric,
					"string data exceeds buffer boundary")
			}
			stringVal := string(buffer[*pos : *pos+int(strLen)])
			v.data = stringVal
			*pos += int(strLen)
		}

	case ValueTypeBytes:
		bufLen := binary.LittleEndian.Uint32(buffer[*pos:])
		*pos += 4

		if bufLen == 0 {
			v.data = ([]byte)(nil)
		} else {
			if *pos+int(bufLen) > len(buffer) {
				return newBufferProtocolError(ErrorCodeGeneric,
					"bytes data exceeds buffer boundary")
			}
			bytesVal := make([]byte, bufLen)
			copy(bytesVal, buffer[*pos:*pos+int(bufLen)])
			v.data = bytesVal
			*pos += int(bufLen)
		}

	case ValueTypeArray:
		arrayLen := binary.LittleEndian.Uint32(buffer[*pos:])
		*pos += 4

		arrayVal := make([]Value, arrayLen)

		for i := uint32(0); i < arrayLen; i++ {
			if *pos >= len(buffer) {
				return newBufferProtocolError(ErrorCodeGeneric,
					"array item exceeds buffer boundary")
			}

			itemType := bufferTypeToValueType(buffer[*pos])
			*pos++

			item := &Value{typ: itemType}
			err := item.deserializeContent(buffer, pos)
			if err != nil {
				return err
			}

			arrayVal[i] = *item
		}
		v.data = arrayVal

	case ValueTypeObject:
		objSize := binary.LittleEndian.Uint32(buffer[*pos:])
		*pos += 4

		objectVal := make(map[string]Value)

		for i := uint32(0); i < objSize; i++ {
			// Read key
			if *pos+4 > len(buffer) {
				return newBufferProtocolError(ErrorCodeGeneric,
					"object key length exceeds buffer boundary")
			}

			keyLen := binary.LittleEndian.Uint32(buffer[*pos:])
			*pos += 4

			if *pos+int(keyLen) > len(buffer) {
				return newBufferProtocolError(ErrorCodeGeneric,
					"object key data exceeds buffer boundary")
			}

			key := string(buffer[*pos : *pos+int(keyLen)])
			*pos += int(keyLen)

			// Read value
			if *pos >= len(buffer) {
				return newBufferProtocolError(ErrorCodeGeneric,
					"object value type exceeds buffer boundary")
			}

			valType := bufferTypeToValueType(buffer[*pos])
			*pos++

			val := &Value{typ: valType}
			err := val.deserializeContent(buffer, pos)
			if err != nil {
				return err
			}

			objectVal[key] = *val
		}
		v.data = objectVal

	default:
		panic("unsupported value type for deserialization")
	}

	return nil
}

File: core/src/ten_runtime/binding/python/interface/ten_runtime/value_buffer.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
import struct
from typing import cast

from .value import Value, ValueType


# Buffer protocol constants - must match C layer
VALUE_BUFFER_MAGIC = 0x010E
VALUE_BUFFER_VERSION = 1
VALUE_BUFFER_HEADER_SIZE = 8  # magic(2) + version(1) + type(1) + size(4)

# Buffer type constants - must match TEN_VALUE_BUFFER_TYPE in C
BUFFER_TYPE_INVALID = 0
BUFFER_TYPE_BOOL = 1
BUFFER_TYPE_INT8 = 2
BUFFER_TYPE_INT16 = 3
BUFFER_TYPE_INT32 = 4
BUFFER_TYPE_INT64 = 5
BUFFER_TYPE_UINT8 = 6
BUFFER_TYPE_UINT16 = 7
BUFFER_TYPE_UINT32 = 8
BUFFER_TYPE_UINT64 = 9
BUFFER_TYPE_FLOAT32 = 10
BUFFER_TYPE_FLOAT64 = 11
BUFFER_TYPE_STRING = 12
BUFFER_TYPE_BUF = 13
BUFFER_TYPE_ARRAY = 14
BUFFER_TYPE_OBJECT = 15
BUFFER_TYPE_PTR = 16
BUFFER_TYPE_JSON_STRING = 17


class ValueBufferHeader:
    """Represents the buffer header structure."""

    magic: int
    version: int
    type_id: int
    size: int

    def __init__(self, magic: int, version: int, type_id: int, size: int):
        self.magic = magic
        self.version = version
        self.type_id = type_id
        self.size = size


def _value_type_to_buffer_type(value_type: ValueType) -> int:
    """Convert Python ValueType to buffer type."""
    mapping = {
        ValueType.INVALID: BUFFER_TYPE_INVALID,
        ValueType.BOOL: BUFFER_TYPE_BOOL,
        ValueType.INT: BUFFER_TYPE_INT64,  # Python int maps to int64
        ValueType.FLOAT: BUFFER_TYPE_FLOAT64,  # Python float maps to float64
        ValueType.STRING: BUFFER_TYPE_STRING,
        ValueType.BYTES: BUFFER_TYPE_BUF,
        ValueType.ARRAY: BUFFER_TYPE_ARRAY,
        ValueType.OBJECT: BUFFER_TYPE_OBJECT,
        ValueType.JSON_STRING: BUFFER_TYPE_JSON_STRING,
    }
    return mapping.get(value_type, BUFFER_TYPE_INVALID)


def _buffer_type_to_value_type(buffer_type: int) -> ValueType:
    """Convert buffer type to Python ValueType."""
    mapping = {
        BUFFER_TYPE_INVALID: ValueType.INVALID,
        BUFFER_TYPE_BOOL: ValueType.BOOL,
        BUFFER_TYPE_INT8: ValueType.INT,
        BUFFER_TYPE_INT16: ValueType.INT,
        BUFFER_TYPE_INT32: ValueType.INT,
        BUFFER_TYPE_INT64: ValueType.INT,
        BUFFER_TYPE_UINT8: ValueType.INT,
        BUFFER_TYPE_UINT16: ValueType.INT,
        BUFFER_TYPE_UINT32: ValueType.INT,
        BUFFER_TYPE_UINT64: ValueType.INT,
        BUFFER_TYPE_FLOAT32: ValueType.FLOAT,
        BUFFER_TYPE_FLOAT64: ValueType.FLOAT,
        BUFFER_TYPE_STRING: ValueType.STRING,
        BUFFER_TYPE_BUF: ValueType.BYTES,
        BUFFER_TYPE_ARRAY: ValueType.ARRAY,
        BUFFER_TYPE_OBJECT: ValueType.OBJECT,
        BUFFER_TYPE_JSON_STRING: ValueType.JSON_STRING,
    }
    return mapping.get(buffer_type, ValueType.INVALID)


def _calculate_content_size(value: Value) -> int:
    """Calculate the size needed for the value content."""
    value_type = value.get_type()

    match value_type:
        case ValueType.INVALID:
            assert False, "Invalid value type"

        case ValueType.BOOL:
            return 1

        case ValueType.INT:
            return 8  # Always serialize as int64

        case ValueType.FLOAT:
            return 8  # Always serialize as float64

        case ValueType.STRING:
            data = value.get_string()[0]
            encoded = data.encode("utf-8")
            return 4 + len(encoded)  # length(4) + data

        case ValueType.JSON_STRING:
            data = value.get_json_string()[0]
            encoded = data.encode("utf-8")
            return 4 + len(encoded)  # length(4) + data

        case ValueType.BYTES:
            data = value.get_buf()[0]
            return 4 + len(data)  # length(4) + data

        case ValueType.ARRAY:
            size = 4  # array length
            for item in value.get_array()[0]:
                size += 1  # item type
                size += _calculate_content_size(item)
            return size

        case ValueType.OBJECT:
            size = 4  # object size
            for key, val in value.get_object()[0].items():
                key_bytes = key.encode("utf-8")
                size += 4 + len(key_bytes)  # key length + key data
                size += 1  # value type
                size += _calculate_content_size(val)
            return size

        case _:  # pyright: ignore[reportUnnecessaryComparison]
            assert (  # pyright: ignore[reportUnreachable]
                False
            ), "Invalid value type"


def _serialize_content(value: Value, buffer: bytearray, pos: int) -> int:
    """Serialize the value content to buffer. Returns new position."""
    value_type = value.get_type()

    match value_type:
        case ValueType.INVALID:
            assert False, "Invalid value type"

        case ValueType.BOOL:
            val = 1 if value.get_bool()[0] else 0
            struct.pack_into("<B", buffer, pos, val)
            pos += 1

        case ValueType.INT:
            # Always serialize as int64
            val = value.get_int()[0]
            struct.pack_into("<q", buffer, pos, val)
            pos += 8

        case ValueType.FLOAT:
            # Always serialize as float64
            val = value.get_float()[0]
            struct.pack_into("<d", buffer, pos, val)
            pos += 8

        case ValueType.STRING:
            data = value.get_string()[0]
            encoded = data.encode("utf-8")
            data_len = len(encoded)
            struct.pack_into("<I", buffer, pos, data_len)
            pos += 4
            if data_len > 0:
                buffer[pos : pos + data_len] = encoded
                pos += data_len

        case ValueType.JSON_STRING:
            data = value.get_json_string()[0]
            encoded = data.encode("utf-8")
            data_len = len(encoded)
            struct.pack_into("<I", buffer, pos, data_len)
            pos += 4
            if data_len > 0:
                buffer[pos : pos + data_len] = encoded
                pos += data_len

        case ValueType.BYTES:
            data = value.get_buf()[0]
            data_len = len(data)
            struct.pack_into("<I", buffer, pos, data_len)
            pos += 4
            if data_len > 0:
                buffer[pos : pos + data_len] = data
                pos += data_len

        case ValueType.ARRAY:
            array_data = value.get_array()[0]
            array_len = len(array_data)
            struct.pack_into("<I", buffer, pos, array_len)
            pos += 4
            for item in array_data:
                item_type = _value_type_to_buffer_type(item.get_type())
                struct.pack_into("<B", buffer, pos, item_type)
                pos += 1
                pos = _serialize_content(item, buffer, pos)

        case ValueType.OBJECT:
            obj_data = value.get_object()[0]
            obj_size = len(obj_data)
            struct.pack_into("<I", buffer, pos, obj_size)
            pos += 4
            for key, val in obj_data.items():
                # Write key
                key_bytes = key.encode("utf-8")
                key_len = len(key_bytes)
                struct.pack_into("<I", buffer, pos, key_len)
                pos += 4
                buffer[pos : pos + key_len] = key_bytes
                pos += key_len
                # Write value type and content
                val_type = _value_type_to_buffer_type(val.get_type())
                struct.pack_into("<B", buffer, pos, val_type)
                pos += 1
                pos = _serialize_content(val, buffer, pos)

        case _:  # pyright: ignore[reportUnnecessaryComparison]
            assert (  # pyright: ignore[reportUnreachable]
                False
            ), f"Unsupported value type for serialization: {value_type}"

    return pos


def serialize_to_buffer(value: Value) -> bytes:
    """Serialize a Value to a buffer using only Python operations."""
    content_size = _calculate_content_size(value)
    total_size = VALUE_BUFFER_HEADER_SIZE + content_size
    buffer = bytearray(total_size)

    pos = 0

    # Write header
    header = ValueBufferHeader(
        magic=VALUE_BUFFER_MAGIC,
        version=VALUE_BUFFER_VERSION,
        type_id=_value_type_to_buffer_type(value.get_type()),
        size=content_size,
    )

    struct.pack_into(
        "<HBBcI",
        buffer,
        pos,
        header.magic,
        header.version,
        header.type_id,
        content_size,
    )
    pos += VALUE_BUFFER_HEADER_SIZE

    # Write content
    final_pos = _serialize_content(value, buffer, pos)

    if final_pos != total_size:
        assert (
            False
        ), f"Buffer size mismatch: expected {total_size}, got {final_pos}"

    return bytes(buffer)


def _validate_buffer_header(buffer: bytes) -> ValueBufferHeader:
    """Validate the buffer header and return header info."""
    if len(buffer) < VALUE_BUFFER_HEADER_SIZE:
        assert False, "Buffer too small to contain header"

    magic, version, type_id, size = cast(
        tuple[int, int, int, int], struct.unpack_from("<HBBIN", buffer, 0)
    )

    if magic != VALUE_BUFFER_MAGIC:
        assert False, "Invalid buffer magic number"

    if version != VALUE_BUFFER_VERSION:
        assert False, "Unsupported buffer protocol version"

    if _buffer_type_to_value_type(type_id) == ValueType.INVALID:
        assert False, "Invalid or unknown buffer type"

    if len(buffer) < VALUE_BUFFER_HEADER_SIZE + size:
        assert False, "Buffer size doesn't match header specification"

    return ValueBufferHeader(magic, version, type_id, size)


def _deserialize_content(
    buffer: bytes, pos: int, value_type: ValueType
) -> tuple[Value, int]:
    """Deserialize value content from buffer. Returns (value, new_position)."""

    match value_type:
        case ValueType.INVALID:
            assert False, "Invalid value type"

        case ValueType.BOOL:
            if pos >= len(buffer):
                assert False, "Buffer too small for bool value"
            val = cast(bool, struct.unpack_from("<B", buffer, pos)[0])
            return Value.from_bool(val != 0), pos + 1

        case ValueType.INT:
            if pos + 8 > len(buffer):
                assert False, "Buffer too small for int value"
            val = cast(int, struct.unpack_from("<q", buffer, pos)[0])
            return Value.from_int(val), pos + 8

        case ValueType.FLOAT:
            if pos + 8 > len(buffer):
                assert False, "Buffer too small for float value"
            val = cast(float, struct.unpack_from("<d", buffer, pos)[0])
            return Value.from_float(val), pos + 8

        case ValueType.STRING | ValueType.JSON_STRING:
            if pos + 4 > len(buffer):
                assert False, "Buffer too small for string length"
            str_len = cast(int, struct.unpack_from("<I", buffer, pos)[0])
            pos += 4

            if str_len == 0:
                data = ""
            else:
                if pos + str_len > len(buffer):
                    assert False, "Buffer too small for string data"
                data_bytes = buffer[pos : pos + str_len]
                data = data_bytes.decode("utf-8")
                pos += str_len

            if value_type == ValueType.STRING:
                return Value.from_string(data), pos
            else:
                return Value.from_json_string(data), pos

        case ValueType.BYTES:
            if pos + 4 > len(buffer):
                assert False, "Buffer too small for bytes length"
            buf_len = cast(int, struct.unpack_from("<I", buffer, pos)[0])
            pos += 4

            if buf_len == 0:
                data = b""
            else:
                if pos + buf_len > len(buffer):
                    assert False, "Buffer too small for bytes data"
                data = bytes(buffer[pos : pos + buf_len])
                pos += buf_len

            return Value.from_buf(data), pos

        case ValueType.ARRAY:
            if pos + 4 > len(buffer):
                assert False, "Buffer too small for array length"
            array_len = cast(int, struct.unpack_from("<I", buffer, pos)[0])
            pos += 4

            array_data: list[Value] = []
            for _ in range(array_len):
                if pos >= len(buffer):
                    assert False, "Buffer too small for array item type"
                item_type_id = cast(
                    int, struct.unpack_from("<B", buffer, pos)[0]
                )
                pos += 1

                item_type = _buffer_type_to_value_type(item_type_id)
                item, pos = _deserialize_content(buffer, pos, item_type)
                array_data.append(item)

            return Value.from_array(array_data), pos

        case ValueType.OBJECT:
            if pos + 4 > len(buffer):
                assert False, "Buffer too small for object size"
            obj_size = cast(int, struct.unpack_from("<I", buffer, pos)[0])
            pos += 4

            obj_data: dict[str, Value] = {}
            for _ in range(obj_size):
                # Read key
                if pos + 4 > len(buffer):
                    assert False, "Buffer too small for object key length"
                key_len = cast(int, struct.unpack_from("<I", buffer, pos)[0])
                pos += 4

                if pos + key_len > len(buffer):
                    assert False, "Buffer too small for object key data"
                key_bytes = buffer[pos : pos + key_len]
                key = key_bytes.decode("utf-8")
                pos += key_len

                # Read value
                if pos >= len(buffer):
                    assert False, "Buffer too small for object value type"
                val_type_id = cast(
                    int, struct.unpack_from("<B", buffer, pos)[0]
                )
                pos += 1

                val_type = _buffer_type_to_value_type(val_type_id)
                val, pos = _deserialize_content(buffer, pos, val_type)
                obj_data[key] = val

            return Value.from_object(obj_data), pos

        case _:  # pyright: ignore[reportUnnecessaryComparison]
            assert (  # pyright: ignore[reportUnreachable]
                False
            ), f"Unknown value type: {value_type}"


def deserialize_from_buffer(buffer: bytes) -> Value:
    """Deserialize a Value from buffer."""
    header = _validate_buffer_header(buffer)

    pos = VALUE_BUFFER_HEADER_SIZE
    value_type = _buffer_type_to_value_type(header.type_id)

    value, _ = _deserialize_content(buffer, pos, value_type)

    return value
